### PR TITLE
[MLIR][Arith] Add rounding mode attribute to `arith.truncf`

### PR DESCRIPTION
**Pull request by victor-eds**
Add rounding mode attributes to `arith`. Use it in `arith.truncf` for now.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Arith/IR/ArithBase.td [hunk 1]
```diff
@@ -156,4 +156,26 @@ def Arith_IntegerOverflowAttr :
   let assemblyFormat = "`<` $value `>`";
 }
 
+//===----------------------------------------------------------------------===//
+// Arith_RoundingMode
+//===----------------------------------------------------------------------===//
+
```
Comment (sommerlukas) [inline mlir/include/mlir/Dialect/Arith/IR/ArithBase.td:162]: Are these modes and their names defined somewhere that we can link here?

File: mlir/include/mlir/Dialect/Arith/IR/ArithOpsInterfaces.td [hunk 2]
```diff
@@ -106,4 +106,26 @@ def ArithIntegerOverflowFlagsInterface : OpInterface<"ArithIntegerOverflowFlagsI
...
+    Access to op rounding mode.
+  }];
+
+  let cppNamespace = "::mlir::arith";
+
+  let methods = [
+    InterfaceMethod<
+      /*desc=*/        "Returns an RoundingModeAttr attribute for the operation",
+      /*returnType=*/  "RoundingModeAttr",
+      /*methodName=*/  "getRoundingModeAttr",
```
Comment (sommerlukas) [inline mlir/include/mlir/Dialect/Arith/IR/ArithOpsInterfaces.td:120]: Do we need the `Attr` suffix on the name?

**Review by sommerlukas — COMMENTED** [review]
Just some nits.

File: mlir/include/mlir/Dialect/Arith/IR/ArithBase.td (hunk 1 above)
Comment (victor-eds) [reply mlir/include/mlir/Dialect/Arith/IR/ArithBase.td:162]: These correspond to LLVM's. I can link.

File: mlir/include/mlir/Dialect/Arith/IR/ArithOpsInterfaces.td (hunk 2 above)
Comment (victor-eds) [reply mlir/include/mlir/Dialect/Arith/IR/ArithOpsInterfaces.td:120]: Yes, as that's returning an attribute. That's MLIR convention.

**Review by sommerlukas — APPROVED** [review]

**Review by etiotto — COMMENTED** [review]
Not sure whether the community will accept this change given that the new operations cannot be lowered to the SPIRV dialect.

File: mlir/include/mlir/Dialect/Arith/IR/ArithBase.td [hunk 3]
```diff
@@ -156,4 +156,29 @@ def Arith_IntegerOverflowAttr :
...
+// These correspond to LLVM's values defined in:
+// llvm/include/llvm/ADT/FloatingPointMode.h
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/Arith/IR/ArithBase.td:164]: Given that the arith dialect operations could be converted to the APIRV dialect these enums need not match the LLVM values.

**Comment by victor-eds** [comment]
> Not sure whether the community will accept this change given that the new operations cannot be lowered to the SPIRV dialect.

They can, but that's missing in the dialect as of now (rounding mode attribute):

```
tonearesteven -> RTE
towardzero -> RTZ
upward -> RTP
downward -> RTN
tonearestaway -> not in SPIR-V
```

File: mlir/include/mlir/Dialect/Arith/IR/ArithBase.td (hunk 3 above)
Comment (victor-eds) [reply mlir/include/mlir/Dialect/Arith/IR/ArithBase.td:164]: They don't exactly match (`tonearesteven` vs `tonearest`)... Maybe we can just drop that claim.

**Review by etiotto — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**